crates/nargo/src/cli/prove_cmd.rs

```diff
@@ -167,27 +61,51 @@ pub fn parse_and_solve_witness<P: AsRef<Path>>(
         noirc_abi::input_parser::Format::Toml.parse(&program_dir, PROVER_INPUT_FILE)?;
 
     // Solve the remaining witnesses
-    let (solved_witness, return_value) = solve_witness(compiled_program, &witness_map)?;
+    let solved_witness = solve_witness(compiled_program, &witness_map)?;
+
+    // We allow the user to optionally not provide a value for the circuit's return value, so this may be missing from
+    // `witness_map`. We must then decode these from the circuit's witness values.
+    let encoded_public_inputs: Vec<FieldElement> = compiled_program
+        .circuit
+        .public_inputs
+        .0
+        .iter()
+        .map(|index| solved_witness[index])
+        .collect();
+
+    let public_abi = compiled_program.abi.as_ref().unwrap().clone().public_abi();
+    let public_inputs = public_abi.decode(&encoded_public_inputs)?;
 
     // Write public inputs into Verifier.toml
-    let abi = compiled_program.abi.as_ref().unwrap();
-    export_public_inputs(return_value, &solved_witness, &witness_map, abi, &program_dir)?;
+    noirc_abi::input_parser::Format::Toml.serialise(
+        &program_dir,
+        VERIFIER_INPUT_FILE,
+        &public_inputs,
+    )?;
 
     Ok(solved_witness)
 }
 
 fn solve_witness(
     compiled_program: &noirc_driver::CompiledProgram,
     witness_map: &BTreeMap<String, InputValue>,
-) -> Result<(BTreeMap<Witness, FieldElement>, Option<Witness>), CliError> {
+) -> Result<BTreeMap<Witness, FieldElement>, CliError> {
     let abi = compiled_program.abi.as_ref().unwrap();
-    let (mut solved_witness, return_value) = process_abi_with_input(abi.clone(), witness_map)
-        .map_err(|error| match error {
-            AbiError::UndefinedInput(_) => {
-                CliError::Generic(format!("{} in the {}.toml file.", error, PROVER_INPUT_FILE))
-            }
-            _ => CliError::from(error),
-        })?;
+    let encoded_inputs = abi.clone().encode(witness_map, true).map_err(|error| match error {
+        AbiError::UndefinedInput(_) => {
+            CliError::Generic(format!("{} in the {}.toml file.", error, PROVER_INPUT_FILE))
+        }
+        _ => CliError::from(error),
+    })?;
+
+    let mut solved_witness: BTreeMap<Witness, FieldElement> = encoded_inputs
+        .into_iter()
+        .enumerate()
+        .map(|(index, witness_value)| {
+            let witness = Witness::new(WITNESS_OFFSET + (index as u32));
+            (witness, witness_value)
+        })
+        .collect();
```

Comment:
Ideally we'd have something like `compiled_program.circuit.inputs` so we could write in the inputs without needing to know about the witness offset similarly to how we're extracting the public inputs.